app/ota_client.py

```diff
@@ -896,48 +887,149 @@ def _gen_regular_files(self, rootfs_dir, regulars_file, target_dir):
+            rfiles_list = [RegularInf(l) for l in f.readlines()]
+
+        # process all regular files here
+        try:
+            self._process_regular_files(rootfs_dir, rfiles_list, target_dir)
+        except:
+            logger.exception(f"gen regular files error!")
+            return False
+        finally:
+            os.chdir(cwd)
+
+        return True
+
+    def _process_regular_files_pool_init(self, gvar_dict):
+        """
+        Used by _process_regular_files
+        Init the worker pool with shared variables
+
+        DO NOT call this method from other functions
+        except for _process_regular_files
+        """
+        global global_var_dict
+        global_var_dict = gvar_dict
+
+    def _process_regular_files_exit(self, gvar_dict):
+        """
+        Used by _process_regular_files
+        Update corresponding class attributes
+
+        DO NOT call this method from other functions
+        except for _process_regular_files
+        """
+        for k, v in gvar_dict.items():
+            # apply staged data to target attribute
+            vname = k.split("-")
+            if vname[0] == "staging":
+                if vname[1] == "dict":
+                    # TODO: update or replace?
+                    setattr(self, vname[-1], dict(v))
+                elif vname[1] == "list":
+                    setattr(self, vname[-1], list(v))
+
+    def _process_regular_files(self, rootfs_dir, rfiles_list, target_dir):
+        with Manager() as manager:
+            ecb_queue = manager.Queue()
+            # variables passed to child processes
+            # variable naming pattern: <prefix>-<type>-<var_name>
+            # prefix tmp: for temporary use
+            # prefix staging: used to update corresponding class attribute
+            gvar_dict = {
+                "tmp-dict-hardlink_reg": manager.dict(),
+                "staging-dict-_rollback_dict": manager.dict(),
+            }
+
+            # default to one worker per CPU core
+            with Pool(
+                initializer=self._process_regular_files_pool_init, initargs=(gvar_dict,)
+            ) as pool:
+
+                # error_callback for workers
+                #   terminate all sub-processes
+                #   and pass exception to main process
+                def ecb(e):
+                    ecb_queue.put(e)
+                    pool.terminate()
+
+                for rfile_inf in rfiles_list:
+                    if (
+                        int(rfile_inf.links) >= 2
+                        and rfile_inf.sha256hash
+                        not in gvar_dict["tmp-dict-hardlink_reg"]
+                    ):
+                        # block the flow until the first copy of hardlinked file is ready
+                        try:
+                            pool.apply(
+                                self._process_regular_file,
+                                (rootfs_dir, target_dir, rfile_inf),
```

Comment:
I think adding `gvar_dict` to the _process_regular_file argument is enough.
Also I think following code is enough:
```suggestion
                       hardlink_dict = manager.dict()
                           for rfile_inf in rfiles_list:
                                ...
                                (rootfs_dir, target_dir, rfile_inf, hardlink_dict),
```

Reply:
Actually both ways are common methods to pass a sharing object to sub processes.

But in this case, since we already use **gvar_dict** to distribute sharing objects, to unify the sharing method, I think we should keep using the sharing objects by **gvar_dict**.